### PR TITLE
reimplement `push_down_projection` and `prune_column`.

### LINKED ISSUE #4265: Reimplement `projection_push_down`
**Describe the bug**

**To Reproduce**

**Expected behavior**

**Additional context**

### LINKED ISSUE #4267: [EPIC]: reimplement all rule which contains global-state
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Current some rule contains `global-state`

Such as `limit_push_down` use `ancestor: &Ancestor` collect the limit-state of whole tree.

- #4266 
- #4265 
- #4264 
- #4263 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #4265
Closes #4267.

# Rationale for this change

original `push_down_projection ` use  `HashSet<Column>` store `required column` and push it through plan top to down.

Now, I remove it and use `Projection` Plan itself to require child to need to output some columns.

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: benchmarks/expected-plans/q19.txt [hunk 1]
```diff
@@ -2,8 +2,13 @@ Projection: SUM(lineitem.l_extendedprice * Int64(1) - lineitem.l_discount) AS re
...
...ome(2000),15,2) AND lineitem.l_quantity <= Decimal128(Some(3000),15,2)lineitem.l_quantity <= Decimal128(Some(3000),15,2)Decimal128(Some(3000),15,2)lineitem.l_quantitylineitem.l_quantity >= Decimal128(Some(2000),15,2)Decimal128(Some(2000),15,2)lineitem.l_quantitylineitem.l_quantity >= Decimal128(Some(100),15,2) AND lineitem.l_quantity <= Decimal128(Some(1100),15,2) OR lineitem.l_quantity >= Decimal128(Some(1000),15,2) AND lineitem.l_quantity <= Decimal128(Some(2000),15,2)lineitem.l_quantity >= Decimal128(Some(1000),15,2) AND lineitem.l_quantity <= Decimal128(Some(2000),15,2)lineitem.l_quantity <= Decimal128(Some(2000),15,2)Decimal128(Some(2000),15,2)lineitem.l_quantitylineitem.l_quantity >= Decimal128(Some(1000),15,2)Decimal128(Some(1000),15,2)lineitem.l_quantitylineitem.l_quantity >= Decimal128(Some(100),15,2) AND lineitem.l_quantity <= Decimal128(Some(1100),15,2)lineitem.l_quantity <= Decimal128(Some(1100),15,2)Decimal128(Some(1100),15,2)lineitem.l_quantitylineitem.l_quantity >= Decimal128(Some(100),15,2)Decimal128(Some(100),15,2)lineitem.l_quantity, lineitem.l_partkey, lineitem.l_quantity, lineitem.l_extendedprice, lineitem.l_discount, lineitem.l_shipinstruct, lineitem.l_shipmode
```
Comment (jackwener) [inline benchmarks/expected-plans/q19.txt:9]: It may be related with #3903?
cc @Ted-Jiang , @alamb .
Look like `Factorize predicate` exist bug.

**Comment by jackwener** [comment]
The current PR does not pass the CI, but it looks like this is due to a bug unrelated to this PR, which we need to fix first.

TPCH-19 will failed in the bug.

File: benchmarks/expected-plans/q10.txt [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
-        Inner Join: orders.o_orderkey = lineitem.l_orderkey
-          Inner Join: customer.c_custkey = orders.o_custkey
-            TableScan: customer projection=[c_custkey, c_name, c_address, c_nationkey, c_phone, c_acctbal, c_comment]
-            Filter: orders.o_orderdate >= Date32("8674") AND orders.o_orderdate < Date32("8766")
-              TableScan: orders projection=[o_orderkey, o_custkey, o_orderdate]
-          Filter: lineitem.l_returnflag = Utf8("R")
-            TableScan: lineitem projection=[l_orderkey, l_extendedprice, l_discount, l_returnflag]
-        TableScan: nation projection=[n_nationkey, n_name]
\ No newline at end of file
+      Projection: customer.c_custkey, customer.c_name, customer.c_address, customer.c_phone, customer.c_acctbal, customer.c_comment, lineitem.l_extendedprice, lineitem.l_discount, nation.n_name
```
Comment (jackwener) [inline benchmarks/expected-plans/q10.txt:4]: We can find some new projection happen above join. because I prune clone for `Join`

File: benchmarks/expected-plans/q19.txt (hunk 1 above)
Comment (alamb) [reply benchmarks/expected-plans/q19.txt:9]: Can you be a little more specific about what the bug is? I would be happy to try and help fix it

Comment (jackwener) [reply benchmarks/expected-plans/q19.txt:9]: I investigate this Problem, it's caused by `common_sub_expression_eliminate`.

tracked in issue #4575

**Comment by alamb** [comment]
Marking as draft to signify that this PR is not quite ready for re-review -- please mark it ready for review when it is ready

**Comment by alamb** [comment]
I plan to help review this over the next day or two

File: datafusion/optimizer/tests/integration-test.rs [hunk 3]
```diff
@@ -195,9 +195,10 @@ fn between_date64_plus_interval() -> Result<()> {
     WHERE col_date64 between '1998-03-18T00:00:00' AND cast('1998-03-18' as date) + INTERVAL '90 days'";
     let plan = test_sql(sql)?;
     let expected =
-        "Projection: COUNT(Int64(1))\n  Aggregate: groupBy=[[]], aggr=[[COUNT(Int64(1))]]\
+        "Aggregate: groupBy=[[]], aggr=[[COUNT(Int64(1))]]\
+        \n  Projection: test.col_int32\
         \n    Filter: test.col_date64 >= Date64(\"890179200000\") AND test.col_date64 <= Date64(\"897955200000\")\
-        \n      TableScan: test projection=[col_date64]";
+        \n      TableScan: test projection=[col_int32, col_date64]";
```
Comment (jackwener) [inline datafusion/optimizer/tests/integration-test.rs:201]: One extra column is cause by `aggr=[[COUNT(Int64(1))]]`.
When aggr don't use any column, pushdown_projection will push first column of child schema.

File: datafusion/optimizer/tests/integration-test.rs [hunk 4]
```diff
@@ -135,10 +135,10 @@ fn anti_join_with_join_filter() -> Result<()> {
...
     let plan = test_sql(sql)?;
     let expected = "Projection: test.col_utf8\
-                    \n  LeftAnti Join: test.col_int32 = t2.col_int32 Filter: test.col_uint32 != t2.col_uint32\
-                    \n    TableScan: test projection=[col_int32, col_uint32, col_utf8]\
-                    \n    SubqueryAlias: t2\
-                    \n      TableScan: test projection=[col_int32, col_uint32, col_utf8]";
+    \n  LeftAnti Join: test.col_int32 = t2.col_int32 Filter: test.col_uint32 != t2.col_uint32\
+    \n    TableScan: test projection=[col_int32, col_uint32, col_utf8]\
+    \n    SubqueryAlias: t2\
+    \n      TableScan: test projection=[col_int32, col_uint32]";
```
Comment (jackwener) [inline datafusion/optimizer/tests/integration-test.rs:141]: The original prune column isn't enough.

**Comment by jackwener** [comment]
cc @liukun4515 @mingmwang

File: benchmarks/expected-plans/q19.txt [hunk 5]
```diff
@@ -1,8 +1,11 @@
...
-          TableScan: lineitem projection=[l_partkey, l_quantity, l_extendedprice, l_discount, l_shipinstruct, l_shipmode]
-        Filter: (part.p_brand = Utf8("Brand#12") AND part.p_container IN ([Utf8("SM CASE"), Utf8("SM BOX"), Utf8("SM PACK"), Utf8("SM PKG")]) AND part.p_size <= Int32(5) OR part.p_brand = Utf8("Brand#23") AND part.p_container IN ([Utf8("MED BAG"), Utf8("MED BOX"), Utf8("MED PKG"), Utf8("MED PACK")]) AND part.p_size <= Int32(10) OR part.p_brand = Utf8("Brand#34") AND part.p_container IN ([Utf8("LG CASE"), Utf8("LG BOX"), Utf8("LG PACK"), Utf8("LG PKG")]) AND part.p_size <= Int32(15)) AND part.p_size >= Int32(1)
-          TableScan: part projection=[p_partkey, p_brand, p_size, p_container]
+    Projection: lineitem.l_extendedprice, lineitem.l_discount
+      Projection: lineitem.l_quantity, lineitem.l_extendedprice, lineitem.l_discount, part.p_brand, part.p_size, part.p_container
```
Comment (jackwener) [inline benchmarks/expected-plans/q19.txt:4]: There are two projections because

```sql
projection
  filter
    projection
```

pushdown filter ->

```sql
projection
  projection
    filter
```

Comment (jackwener) [inline benchmarks/expected-plans/q19.txt:4]: There are two projections because

```sql
projection
  filter
    projection
```

pushdown filter ->

```sql
projection
  projection
    filter
```

File: datafusion/optimizer/src/optimizer.rs [hunk 6]
```diff
@@ -216,23 +216,24 @@ impl Optimizer {
             // subqueries to joins
             Arc::new(SimplifyExpressions::new()),
             Arc::new(RewriteDisjunctivePredicate::new()),
+            Arc::new(EliminateLimit::new()),
             Arc::new(EliminateFilter::new()),
+            // Filters can't be pushed down past Limits, we should do PushDownFilter after PushDownLimit
```
Comment (liukun4515) [inline datafusion/optimizer/src/optimizer.rs:221]: are there dependencies between some optimization rules?

I don't why we use the same rule twice, such as `PushDownLimit`

**Comment by liukun4515** [comment]
I will take a look this tomorrow.

**Review by alamb — COMMENTED** [review]
This looks epic @jackwener  -- I have this on my list to review tomorrow

File: datafusion/core/tests/sql/avro.rs [hunk 7]
```diff
@@ -140,18 +140,16 @@ async fn avro_explain() {
...
         ],
         vec![
             "physical_plan",
-            "ProjectionExec: expr=[COUNT(UInt8(1))@0 as COUNT(UInt8(1))]\
-            \n  AggregateExec: mode=Final, gby=[], aggr=[COUNT(UInt8(1))]\
-            \n    CoalescePartitionsExec\
-            \n      AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]\
-            \n        RepartitionExec: partitioning=RoundRobinBatch(NUM_CORES), input_partitions=1\
-            \n          AvroExec: files={1 group: [[ARROW_TEST_DATA/avro/alltypes_plain.avro]]}, limit=None\
+            "AggregateExec: mode=Final, gby=[], aggr=[COUNT(UInt8(1))]\
```
Comment (alamb) [inline datafusion/core/tests/sql/avro.rs:148]: 👍  that looks nicer

File: datafusion/core/tests/sql/parquet.rs [hunk 8]
```diff
@@ -309,11 +309,11 @@ async fn parquet_query_with_max_min() {
     let sql = "SELECT max(c1) FROM foo";
     let actual = execute_to_batches(&ctx, sql).await;
     let expected = vec![
-        "+-------------+",
-        "| MAX(foo.c1) |",
-        "+-------------+",
-        "| 3           |",
-        "+-------------+",
+        "+---------+",
+        "| MAX(c1) |",
```
Comment (alamb) [inline datafusion/core/tests/sql/parquet.rs:313]: I am not sure what is responsible for removing the relation name `foo` but I think it is a good change (and more consistent with postgres)

File: datafusion/core/tests/sql/window.rs [hunk 9]
```diff
@@ -2287,13 +2286,17 @@ async fn test_remove_unnecessary_sort_in_sub_query() -> Result<()> {
...
             "      AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]",
             "        RepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=8",
-            "          AggregateExec: mode=FinalPartitioned, gby=[c1@0 as c1], aggr=[COUNT(UInt8(1))]",
-            "            CoalesceBatchesExec: target_batch_size=4096",
-            "              RepartitionExec: partitioning=Hash([Column { name: \"c1\", index: 0 }], 8), input_partitions=8",
-            "                AggregateExec: mode=Partial, gby=[c1@0 as c1], aggr=[COUNT(UInt8(1))]",
-            "                  CoalesceBatchesExec: target_batch_size=4096",
-            "                    FilterExec: c13@1 != C2GT5KVyOPZpgKVl110TyZO0NcJ434",
-            "                      RepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=1",
+            "          ProjectionExec: expr=[COUNT(UInt8(1))@0 as COUNT(UInt8(1))]",
```
Comment (alamb) [inline datafusion/core/tests/sql/window.rs:2289]: I don't understand why there are multiple new ProjectionExec's in this plan

File: datafusion/optimizer/tests/integration-test.rs (hunk 3 above)
Comment (alamb) [reply datafusion/optimizer/tests/integration-test.rs:201]: would it be possible to not push a column down if there was already a column that was scanned (e.g. in this case use col_date64?)

File: datafusion/optimizer/src/push_down_projection.rs [hunk 10]
```diff
@@ -45,445 +42,482 @@ use std::{
...
+                    .map(|(i, e)| {
+                        match e {
+                            Ok(e) => {
+                                let parent_expr = projection
+                                    .schema
+                                    .fields()
+                                    .get(i)
+                                    .unwrap()
+                                    .qualified_name();
+                                // TODO
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_projection.rs:134]: I wonder what this TODO is tracking

**Review by alamb — COMMENTED** [review]
Thank you @jackwener -- I reviewed the code and plan changes carefully

I found the addition of so many `Projection` into the plans confusing as they obscure the key operations somewhat. It seems like the new Projection's are primarily introduced because you  need some way to encode the currently used set of columns now that the recursion happens outside of the optimizer. Is that true?

It would be great to reduce the newly added ProjectionExec somehow. For example, can we introduce a pass that removes any that don't actually reduce the schema meaningfully (eg the input to an aggregate, for example)?

Basically I think this is a very nice change, and a great end to a long epic of work. Thank you so much for your contribution

cc @ygf11 and @liukun4515 who I think have been working on joins recently and may have some more thoughts on this

File: benchmarks/expected-plans/q10.txt (hunk 2 above)
Comment (jackwener) [reply benchmarks/expected-plans/q10.txt:4]: Yes, the `Projection` nodes above the `join` is used to just get columns that we need to use.

Comment (ygf11) [reply benchmarks/expected-plans/q10.txt:4]: Maybe it is relative to [`build_join_schema`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/logical_plan/builder.rs#L935). Some optimizer rules call this function.
I think it is ok for calling it before pushdown projection, but I guess it is not correct after push down projection.
  
For the query:
```sql
select a.id from a join b on a.id = b.id
``` 

`schema(a)`: a.id
`schema(b)`: b.id

`build_join_schema` will merge left and right, the result is `a.id` + `b.id`, but the expected result should be only `a.id`.

Maybe we can fix it first, and then we will not need the projection any more.

Comment (jackwener) [reply benchmarks/expected-plans/q10.txt:4]: I'm sorry for this PR is easily misunderstood. I have corrected it.

**Comment by alamb** [comment]
This PR now sadly appears to have a bunch of conflicts. @jackwener is your idea that with https://github.com/apache/arrow-datafusion/pull/5366 the extra projections introduced by this pass will be removed in the final plans?

**Comment by jackwener** [comment]
> This PR now sadly appears to have a bunch of conflicts. @jackwener is your idea that with #5366 the extra projections introduced by this pass will be removed in the final plans?

I split one part (eliminate projection) of this PR into one new rule, This make rule keep it as simple as possible, don't do too many things at once in one rule. And this PR will be simple.

**Comment by alamb** [comment]
> Yes, this is one of the purposes. As you said above, we can make PR https://github.com/apache/arrow-datafusion/pull/5366 to eliminate the redundant projections that appear in this PR.

Awesome -- I think as long a the redundant projections are removed by the final plan it is fine to introduce them in an earlier pass 👍 

> Another purpose is to split one part (eliminate projection) of this PR into one new rule. This make rule keep it as simple as possible, don't do too many things at once in one rule. And this PR will be more simple than now.

❤️

**Comment by jackwener** [comment]
@ygf11 @alamb @liukun4515 PTAL.

I resolve all problem in this PR. now this PR just add some extra projection. This projection is used for prune-column. 

We can see `Spark` plan to see the expected effect, like [spark tpch q2 plan](https://github.com/apache/spark/blob/master/sql/core/src/test/resources/tpch-plan-stability/q2/explain.txt), you can see there are projection above join or filter.

Because these projection can make we just get columns that we use, prune-column will make `datafusion` use less memory!

File: datafusion/optimizer/tests/integration-test.rs (hunk 3 above)
Comment (jackwener) [reply datafusion/optimizer/tests/integration-test.rs]: > would it be possible to not push a column down if there was already a column that was scanned (e.g. in this case use col_date64?)

I have resolved it !🚀

File: benchmarks/expected-plans/q11.txt [hunk 11]
```diff
@@ -3,18 +3,24 @@ Sort: value DESC NULLS FIRST
...
-            Aggregate: groupBy=[[]], aggr=[[SUM(CAST(partsupp.ps_supplycost AS Decimal128(26, 2)) * CAST(partsupp.ps_availqty AS Decimal128(26, 2)))]]
-              Inner Join: supplier.s_nationkey = nation.n_nationkey
+          Projection: partsupp.ps_partkey, partsupp.ps_availqty, partsupp.ps_supplycost
+            Inner Join: supplier.s_nationkey = nation.n_nationkey
+              Projection: partsupp.ps_partkey, partsupp.ps_availqty, partsupp.ps_supplycost, supplier.s_nationkey
                 Inner Join: partsupp.ps_suppkey = supplier.s_suppkey
-                  TableScan: partsupp projection=[ps_suppkey, ps_availqty, ps_supplycost]
+                  TableScan: partsupp projection=[ps_partkey, ps_suppkey, ps_availqty, ps_supplycost]
                   TableScan: supplier projection=[s_suppkey, s_nationkey]
+              Projection: nation.n_nationkey
```
Comment (jackwener) [inline benchmarks/expected-plans/q11.txt:12]: Here Filter will output two columns, but we just need one column `nation.n_nationkey`, so add projection to prune column

File: datafusion/core/tests/sql/window.rs [hunk 12]
```diff
@@ -1515,13 +1515,14 @@ async fn test_remove_unnecessary_sort_in_sub_query() -> Result<()> {
             "    CoalescePartitionsExec",
             "      AggregateExec: mode=Partial, gby=[], aggr=[COUNT(UInt8(1))]",
             "        RepartitionExec: partitioning=RoundRobinBatch(8), input_partitions=8",
-            "          AggregateExec: mode=FinalPartitioned, gby=[c1@0 as c1], aggr=[COUNT(UInt8(1))]",
+            "          AggregateExec: mode=FinalPartitioned, gby=[c1@0 as c1], aggr=[]",
```
Comment (alamb) [inline datafusion/core/tests/sql/window.rs:1518]: I don't understand this change -- why is there no aggregate anymore?

File: datafusion/optimizer/src/push_down_projection.rs [hunk 13]
```diff
@@ -710,12 +679,11 @@ mod tests {
         let expected = "Projection: test.a, test.c, test.b\
         \n  Filter: test.a > Int32(1)\
         \n    Filter: test.b > Int32(1)\
-        \n      Filter: test.c > Int32(1)\
-        \n        TableScan: test projection=[a, b, c]";
-
-        assert_optimized_plan_eq(&plan, expected);
-
-        Ok(())
+        \n      Projection: test.c, test.a, test.b\
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_projection.rs:682]: These projections above a Filter that just pass the input to the output are also unecessary, right? Maybe we can add a rule to the remove unnecessary projections for these as well (if the schema of the projection's input is the same as the schema of its output)

**Review by alamb — APPROVED** [review]
Looks good to me @jackwener  -- thank you. An epic sequence of pull requests

I think the question of where the `COUNT` aggregate went is probably important to answer before merging.

Also, I think we can make the plans even better (maybe as a follow on PR) by avoiding even more redundant Projections

File: datafusion/optimizer/src/push_down_projection.rs (hunk 13 above)
Comment (jackwener) [reply datafusion/optimizer/src/push_down_projection.rs:682]: Thank you @alamb . Agree with it.
But In this example, the schema of the projection' !=  the schema of its child, because the order already changed. 
In the future, we indeed remove these projection (if it just change order, and it isn't in the top of plan tree, which means that there must be a plannode above this projection that can determine the output schema like agg, other projection ....)

File: datafusion/core/tests/sql/window.rs (hunk 12 above)
Comment (jackwener) [reply datafusion/core/tests/sql/window.rs:1518]: Has resolved it , thanks @alamb

File: datafusion/optimizer/src/push_down_projection.rs (hunk 13 above)
Comment (alamb) [reply datafusion/optimizer/src/push_down_projection.rs:698]: >  because the order already changed.
> In the future, we indeed can remove these projection (if it just change order, and it isn't in the top of plan tree, which means that there must be a plannode above this projection that can

That is an excellent point

> We can enhance EliminateProject Rule

Good idea 👍

**Review by alamb — APPROVED** [review]
Let's do it. Thanks @jackwener
